### ISSUE TITLE
Removing OData Dependencies

### ISSUE THREAD

**Issue by chadwackerman** [Azure/azure-storage-net#372]
I have a 25 line .NET app that copies files to blob storage. It is part of a data cluster.

Imagine my surprise to see the resulting package include separate DLLs in German, French, Spanish, Italian, Japanese, Korean, Russian... plus a few other acronyms I'm not going to look up on Wikipedia.

Amazon Web Services provided full .NET core support for all services in September 2015. It is December 2016 and Microsoft hasn't managed to ship a comparable C# client for the smaller Azure ecosystem. 

I will relink the original issue from 2014 to save us all some time:
https://github.com/Azure/azure-storage-net/issues/97

Should've fixed it then, definitely time to bite the bullet now. Trying to copy a stream to a blob shouldn't require four hours of StackOverflow research, three beers and two readings of outdated roadmap blog posts, or "imports": [ "dnxcore50", "portable-net45+win8" ].

**Comment by mirobers** [comment]
It's not a foolproof process for sure. We're planning on removing the OData dependency, which will completely resolve this. Plus we are about to release a fix which makes the OData DLLs completely optional for those who don't consume the table service. We've supported NetCore from the very beginning so I sympathize with your frustrations at the friction in the process.

We can use this issue to track the removal of OData dependencies.

**Comment by raRaRa** [comment]
Is there any rough timeline for removing the OData dependencies completely? I want to port my web app to .NET core and continue to use table services.

**Comment by pemari-msft** [comment]
@raRaRa We are working on this, but to be clear, we do support Table service using .NET core today. So you should not be blocked on it. /cc @erezvani1529

**Comment by bitbonk** [comment]
If you completely remove odata, what are you going to replace it with?

**Comment by chadwackerman** [comment]
Presumably OData functionality would be moved into a separate package. Or if they're washing their hands of it, handed off to the community for active mismanagement versus Microsoft's passive mismanagement of the technology.

**Comment by mirobers** [comment]
We're implementing the subset we need, similar to the other storage SDKs. This will allow goodness such as true async for OData payloads.

**Comment by erezvani1529** [comment]
Some good news regarding netstandard friendly OData package:

As a result of our collaboration with CoreFx team, we have updated our library ([v8.1](https://www.nuget.org/packages/WindowsAzure.Storage/8.1.0)) to use the latest Netstandard friendly ODataLib packages to eliminate the need to import or add additional TFs. 

Thanks for your patience and help while we are working on the remaining issues with ODataLib.

-Elham

**Comment by raRaRa** [comment]
@erezvani1529 I can't seem to install `WindowsAzure.Storage` to a .NET Standard 1.6 class library. It complains that:
`Package System.Linq.Queryable 4.0.0 is not compatible with netstandard1.6 (.NETStandard,Version=v1.6). `

However, `System.Linq.Queryable` version 4.3.0 seems to support .NET Standard.

Am I doing something wrong, or is this by design in v8.1. It works fine if I install `WindowsAzure.Storage` to a netcore class library since `System.Linq.Queryable` 4.0.0 supports it.

Thanks.

**Comment by erezvani1529** [comment]
Hi @raRaRa,

Thanks very much for reporting this. The issue seems to be happening only with latest VS tooling and project templates. Are you using VS2017 and the latest tooling?

BTW, would you please open a new issue to track this problem? I will update with the latest info I get from CoreFX and VS team to resolve this issue as soon as possible.

Thanks,
Elham

**Comment by raRaRa** [comment]
Hey @erezvani1529 

You're right, I'm using the latest version of VS2017 and I created a fresh class library using .NET Standard. So I assume I'm using the latest tooling.

Absolutely, I've opened a new issue for this here:
https://github.com/Azure/azure-storage-net/issues/411

Thanks for looking into this so quickly. Much appreciated.

**Comment by raRaRa** [comment]
Any update on this? My app can't handle any load because OData isn't truly async.

Thanks.

**Comment by RonakThakkar** [comment]
Good to know that you are removing OData libraries and its dependancies.

I really wonder if OData works well for projects. We removed Microsoft OData libraries from all our projects due to performance issues. I did a simple performance test for some medium to complex query and OData is very slow.

**Comment by Jehoel** [comment]
Does this also include the following additional dependencies?

- `Microsoft.Data.Edm`
- `Microsoft.Data.Services.Client`
- `System.Dynamic.Runtime`
- `System.Spatial`
- `System.Net.Requests`

Granted - this isn't an issue with `WindowsAzure.Storage` specifically - just the awkward way that the NuGet Package Manager handles 2nd-degree dependencies: in a giant flat list. I wish NuGet would display packages as a dependency tree and let me hide packages that I'm not referencing directly, and also giving me the option to exclude a 2nd-degree (or 3rd-degree or higher) dependency from the build output when I know-for-a-fact those assemblies are not used at all - this would eliminate the need for this specific issue so then `WindowsAzure.Storage` can continue to include OData support while allowing those not using it to opt-out of including it.

**Comment by adminnz** [comment]
Nearly a year and still no indication on any actual progress being made on this?

**Comment by erezvani1529** [comment]
Hi @adminnz,

We have removed this dependency already (please checkout the public branch dev-split). There will be announcement regarding this change and CosmosD

**Comment by erezvani1529** [comment]
Hi @adminnz,

We have removed this dependency already (please checkout the public branch dev-split). There will be announcement regarding this change and CosmosDB team table SDK very soon.

Thanks,
Elham

**Comment by erezvani1529** [comment]
Thanks everyone for your patience!

Our new service version finally rolled out and unblocked our release for [v9.0.0](https://www.nuget.org/packages/WindowsAzure.Storage/9.0.0). The package is live now and the OData dependencies has been removed.

Once more thank you for your feedbacks.

-Elham

**Comment by PureKrome** [comment]
![](https://78.media.tumblr.com/43e63461d1e3f22a49b18dbf15227a1d/tumblr_inline_n3t10oQfIh1solpjm.gif)
